Readme.md
NProgress
=========

[![Status](https://api.travis-ci.org/rstacruz/nprogress.svg?branch=master)](http://travis-ci.org/rstacruz/nprogress)
[![npm version](https://img.shields.io/npm/v/nprogress.png)](https://npmjs.org/package/nprogress "View this project on npm")
[![jsDelivr Hits](https://data.jsdelivr.com/v1/package/npm/nprogress/badge?style=rounded)](https://www.jsdelivr.com/package/npm/nprogress)

> Minimalist progress bar

Slim progress bars for Ajax'y applications. Inspired by Google, YouTube, and
Medium.

Installation
------------

Add [nprogress.js] and [nprogress.css] to your project.

```html
<script src='nprogress.js'></script>
<link rel='stylesheet' href='nprogress.css'/>
```

NProgress is available via [bower] and [npm].

    $ npm install --save nprogress

Also available via [unpkg] CDN:

- https://unpkg.com/nprogress@0.2.0/nprogress.js
- https://unpkg.com/nprogress@0.2.0/nprogress.css

[bower]: http://bower.io/search/?q=nprogress
[npm]: https://www.npmjs.org/package/nprogress
[unpkg]: https://unpkg.com/

Basic usage
-----------

Simply call `start()` and `done()` to control the progress bar.

~~~ js
NProgress.start();
NProgress.done();
~~~

### Turbolinks (version 5+)
Ensure you're using Turbolinks 5+, and use
this: (explained [here](https://github.com/rstacruz/nprogress/issues/8#issuecomment-239107109))

~~~ js
$(document).on('turbolinks:click', function() {
  NProgress.start();
});
$(document).on('turbolinks:render', function() {
  NProgress.done();
  NProgress.remove();
});
~~~

### Turbolinks (version 3 and below)
Ensure you're using Turbolinks 1.3.0+, and use
this: (explained [here](https://github.com/rstacruz/nprogress/issues/8#issuecomment-23010560))

~~~ js
$(document).on('page:fetch',   function() { NProgress.start(); });
$(document).on('page:change',  function() { NProgress.done(); });
$(document).on('page:restore', function() { NProgress.remove(); });
~~~

### Pjax
Try this: (explained [here](https://github.com/rstacruz/nprogress/issues/22#issuecomment-36540472))

~~~ js
$(document).on('pjax:start', function() { NProgress.start(); });
$(document).on('pjax:end',   function() { NProgress.done();  });
~~~

Ideas
-----

 * Add progress to your Ajax calls! Bind it to the jQuery `ajaxStart` and
 `ajaxStop` events.

 * Make a fancy loading bar even without Turbolinks/Pjax! Bind it to
 `$(document).ready` and `$(window).load`.

Advanced usage
--------------

__Percentages:__ To set a progress percentage, call `.set(n)`, where *n* is a
number between `0..1`.

~~~ js
NProgress.set(0.0);     // Sorta same as .start()
NProgress.set(0.4);
NProgress.set(1.0);     // Sorta same as .done()
~~~

__Incrementing:__ To increment the progress bar, just use `.inc()`. This
increments it with a random amount. This will never get to 100%: use it for
every image load (or similar).

~~~ js
NProgress.inc();
~~~

If you want to increment by a specific value, you can pass that as a parameter:

~~~ js
NProgress.inc(0.2);    // This will get the current status value and adds 0.2 until status is 0.994
~~~

__Force-done:__ By passing `true` to `done()`, it will show the progress bar
even if it's not being shown. (The default behavior is that *.done()* will not
    do anything if *.start()* isn't called)

~~~ js
NProgress.done(true);
~~~

__Get the status value:__ To get the status value, use `.status`

Configuration
-------------

#### `trickle`
Turn off the automatic incrementing behavior by setting this to `false`. (default: `true`)

~~~ js
NProgress.configure({ trickle: false });
~~~

#### `trickleSpeed`
Adjust how often to trickle/increment, in ms.

~~~ js
NProgress.configure({ trickleSpeed: 200 });
~~~

Customization
-------------

Just edit `nprogress.css` to your liking. Tip: you probably only want to find
and replace occurrences of `#29d`.

The included CSS file is pretty minimal... in fact, feel free to scrap it and
make your own!

Resources
---------

 * [New UI Pattern: Website Loading Bars](http://www.usabilitypost.com/2013/08/19/new-ui-pattern-website-loading-bars/) (usabilitypost.com)

Support
-------

__Bugs and requests__: submit them through the project's issues tracker.<br>
[![Issues](http://img.shields.io/github/issues/rstacruz/nprogress.svg)]( https://github.com/rstacruz/nprogress/issues )

__Questions__: ask them at StackOverflow with the tag *nprogress*.<br>
[![StackOverflow](http://img.shields.io/badge/stackoverflow-nprogress-brightgreen.svg)]( http://stackoverflow.com/questions/tagged/nprogress )

__Chat__: join us at gitter.im.<br>
[![Chat](http://img.shields.io/badge/gitter-rstacruz/nprogress-brightgreen.svg)]( https://gitter.im/rstacruz/nprogress )

[default template]: https://github.com/rstacruz/nprogress/blob/master/nprogress.js#L31
[Turbolinks]: https://github.com/rails/turbolinks
[nprogress.js]: http://ricostacruz.com/nprogress/nprogress.js
[nprogress.css]: http://ricostacruz.com/nprogress/nprogress.css

Thanks
------

**NProgress** © 2013-2017, Rico Sta. Cruz. Released under the [MIT License].<br>
Authored and maintained by Rico Sta. Cruz with help from [contributors].

> [ricostacruz.com](http://ricostacruz.com) &nbsp;&middot;&nbsp;
> GitHub [@rstacruz](https://github.com/rstacruz) &nbsp;&middot;&nbsp;
> Twitter [@rstacruz](https://twitter.com/rstacruz)

[MIT License]: http://mit-license.org/
[contributors]: http://github.com/rstacruz/nprogress/contributors

[![](https://img.shields.io/github/followers/rstacruz.svg?style=social&label=@rstacruz)](https://github.com/rstacruz) &nbsp;
[![](https://img.shields.io/twitter/follow/rstacruz.svg?style=social&label=@rstacruz)](https://twitter.com/rstacruz)


History.md
## Changes by Michael Tamm
 * Removed support for specifying a different parent container
 * Removed spinner
 * Removed function NProgress.promise(...)
 * Removed UMD support
 * Removed NProgress.version field
 * Removed settings and NProgress.configure(...) function
 * Added build script, which produces nprogress.min.js
 * The CSS class 'nprogress-busy' is no longer added to the document
 * Made NProgress.status private

## [v0.2.0] - May 13, 2015

This is a maintenance release with cleanups. No functionality changes were introduced.

 * [#93] - Remove jQuery dependency in component ([@slang800])
 * [#120] - Fix the Readme showing a wrong example for `ease` ([@teeceepee])

Minor changes:

 * [#92] - Fix typo in Readme ([@FND])
 * [#90] - Fix typo in Readme ([@mdxs])
 * [#119] - Fix typo in Readme ([@infertux])
 * [#53] - Use `===` instead of `==` ([@drodil])
 * Update test dependencies

## [v0.1.6] - June 25, 2014

 * [#86] - Add support for specifying a different parent container. ([@jonjaques], [#14], [#33], [#39])
 * [#64] - Fix Require.js support ([#75], [#85])
 * [#54] - Fix component support

## [v0.1.5] - June 21, 2014

 * [#82] - Release properly to the npm registry

## [v0.1.4] - June 21, 2014

 * [#65] - Bower: don't download package.json. Fixes browserify + jquery problem. ([@amelon])
 * [#75] - Fix compatibility with Karma. ([@shaqq])

Internal changes:

 * [#77] - Use SVG to display Travis-CI badge. ([@Mithgol])
 * [#34] - Readme: update .inc() and .status docs ([@lacivert])
 * [#73] - Readme: update year ([@rwholmes])

## [v0.1.3] - March 26, 2014

 * [#7] - Remove jQuery dependency ([#28], [#17], [@rurjur])
 * [#73] - Update Readme to change year to 2014. ([@rwholmes])

## [v0.1.2] - August 21, 2013

Minor update for proper [Bower] and [Component] support.

 * Add Bower support.
 * Fix Component support and use `component/jquery` as a dependency.

## [v0.1.1] - August 21, 2013

Minor fixes.

 * Removed the busy cursor that occurs when loading.
 * Added support for IE7 to IE9. ([#3], [@markbao])
 * Implement `trickleRate` and `trickleSpeed` options.
 * [#5] - Implement the `showSpinner` option to allow removing the spinner. ([#9], [@rahulcs])
 * Registered as a Component in Component.io.
 * [#8] - Updated the Readme with better Turbolinks instructions.

## v0.1.0 - August 20, 2013

Initial release.

[Bower]: http://bower.io
[Component]: http://component.io
[#119]: https://github.com/rstacruz/nprogress/issues/119
[#120]: https://github.com/rstacruz/nprogress/issues/120
[#14]: https://github.com/rstacruz/nprogress/issues/14
[#17]: https://github.com/rstacruz/nprogress/issues/17
[#28]: https://github.com/rstacruz/nprogress/issues/28
[#33]: https://github.com/rstacruz/nprogress/issues/33
[#34]: https://github.com/rstacruz/nprogress/issues/34
[#39]: https://github.com/rstacruz/nprogress/issues/39
[#3]: https://github.com/rstacruz/nprogress/issues/3
[#54]: https://github.com/rstacruz/nprogress/issues/84
[#5]: https://github.com/rstacruz/nprogress/issues/5
[#64]: https://github.com/rstacruz/nprogress/issues/64
[#65]: https://github.com/rstacruz/nprogress/issues/65
[#73]: https://github.com/rstacruz/nprogress/issues/73
[#75]: https://github.com/rstacruz/nprogress/issues/75
[#77]: https://github.com/rstacruz/nprogress/issues/77
[#7]: https://github.com/rstacruz/nprogress/issues/7
[#82]: https://github.com/rstacruz/nprogress/issues/82
[#84]: https://github.com/rstacruz/nprogress/issues/84
[#85]: https://github.com/rstacruz/nprogress/issues/85
[#86]: https://github.com/rstacruz/nprogress/issues/86
[#8]: https://github.com/rstacruz/nprogress/issues/8
[#90]: https://github.com/rstacruz/nprogress/issues/90
[#92]: https://github.com/rstacruz/nprogress/issues/92
[#93]: https://github.com/rstacruz/nprogress/issues/93
[#9]: https://github.com/rstacruz/nprogress/issues/9
[@slang800]: https://github.com/slang800
[@teeceepee]: https://github.com/teeceepee
[@FND]: https://github.com/FND
[@mdxs]: https://github.com/mdxs
[@infertux]: https://github.com/infertux
[@jonjaques]: https://github.com/jonjaques
[@amelon]: https://github.com/amelon
[@shaqq]: https://github.com/shaqq
[@Mithgol]: https://github.com/Mithgol
[@lacivert]: https://github.com/lacivert
[@rwholmes]: https://github.com/rwholmes
[@rurjur]: https://github.com/rurjur
[@markbao]: https://github.com/markbao
[@rahulcs]: https://github.com/rahulcs
[v0.1.6]: https://github.com/rstacruz/nprogress/compare/v0.1.5...v0.1.6
[v0.1.5]: https://github.com/rstacruz/nprogress/compare/v0.1.4...v0.1.5
[v0.1.4]: https://github.com/rstacruz/nprogress/compare/v0.1.3...v0.1.4
[v0.1.3]: https://github.com/rstacruz/nprogress/compare/v0.1.2...v0.1.3
[v0.1.2]: https://github.com/rstacruz/nprogress/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/rstacruz/nprogress/compare/v0.1.0...v0.1.1
[#53]: https://github.com/rstacruz/nprogress/issues/53
[v0.2.0]: https://github.com/rstacruz/nprogress/compare/v0.1.6...v0.2.0
[@drodil]: https://github.com/drodil


test/test.js
(function() {

  var assert = window.chai.assert;

  describe('NProgress', function() {
    var $, NProgress;

    beforeEach(function() {
      $ = window.jQuery;
      NProgress = window.NProgress;
      this.settings = $.extend({}, NProgress.settings);
    });

    afterEach(function() {
      $("#nprogress").remove();
      $('html').attr('class', '');
      NProgress.status = null;

      // Restore settings
      $.extend(NProgress.settings, this.settings);
    });

    describe('.set()', function() {
      it('.set(0) must render', function(done) {
        NProgress.set(0);
        assert.equal($("#nprogress").length, 1);
        assert.equal($("#nprogress .bar").length, 1);
        assert.equal($("#nprogress .peg").length, 1);
        done();
      });

      it('.set(1) should appear and disappear', function(done) {
        NProgress.set(0).set(1);
        assert.equal($("#nprogress").length, 1);

        setTimeout(function() {
          assert.equal($("#nprogress").length, 0);
          done();
        }, 500);
      });

      it('must respect minimum', function() {
        NProgress.set(0);
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('must clamp to minimum', function() {
        NProgress.set(-100);
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('must clamp to maximum', function() {
        NProgress.set(456);
        assert.equal(NProgress.status, null);
      });
    });

    // ----

    describe('.start()', function() {
      it('must render', function(done) {
        NProgress.start();
        assert.equal($("#nprogress").length, 1);
        done();
      });

      it('must respect minimum', function() {
        NProgress.start();
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });
    });

    // ----

    describe('.done()', function() {
      it('must not render without start', function(done) {
        NProgress.done();
        assert.equal($("#nprogress").length, 0);
        done();
      });

      it('.done(true) must render', function(done) {
        NProgress.done(true);
        assert.equal($("#nprogress").length, 1);
        done();
      });
    });

    // ----

    describe('.remove()', function() {
      it('should be removed from <body>', function() {
        NProgress.set(1);
        NProgress.remove();

        var $body = $(document.body);
        assert.equal($body.find('#nprogress').length, 0);
      });
    });

    // ----

    describe('.inc()', function() {
      it('should render', function() {
        NProgress.inc();
        assert.equal($("#nprogress").length, 1);
      });

      it('should start with minimum', function() {
        NProgress.inc();
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('should increment', function() {
        NProgress.start();
        var start = NProgress.status;

        NProgress.inc();
        assert.operator(NProgress.status, '>', start);
      });

      it('should never reach 1.0', function() {
        for (var i=0; i<100; ++i) { NProgress.inc(); }
        assert.operator(NProgress.status, '<', 1.0);
      });
    });

    // -----

    describe('.configure()', function() {
      it('should work', function() {
        NProgress.configure({ minimum: 0.5 });
        assert.equal(NProgress.settings.minimum, 0.5);
      });
    });

  });

})();


nprogress.js
/* NProgress, (c) 2013, 2014 Rico Sta. Cruz - http://ricostacruz.com/nprogress
 * @license MIT */

window.NProgress = (function() {
  var NProgress = {};
  var status = null;
  var positionUsing = '';

  /**
   * Sets the progress bar status, where `n` is a number from `0.0` to `1.0`.
   *
   *     NProgress.set(0.4);
   *     NProgress.set(1.0);
   */

  NProgress.set = function(n) {
    var started = NProgress.isStarted();

    n = clamp(n, 0.08, 1);
    status = (n === 1 ? null : n);

    var progress = NProgress.render(!started),
        bar      = progress.querySelector('#nprogress .bar'),
        speed    = 200; /* ms */

    progress.offsetWidth; /* Repaint */

    queue(function(next) {
      // Set positionUsing if it hasn't already been set
      if (positionUsing === '') positionUsing = getPositioningCSS();

      // Add transition
      css(bar, barPositionCSS(n, speed));

      if (n === 1) {
        // Fade out
        css(progress, {
          transition: 'none',
          opacity: 1
        });
        progress.offsetWidth; /* Repaint */

        setTimeout(function() {
          css(progress, {
            transition: 'all ' + speed + 'ms linear',
            opacity: 0
          });
          setTimeout(function() {
            NProgress.remove();
            next();
          }, speed);
        }, speed);
      } else {
        setTimeout(next, speed);
      }
    });

    return this;
  };

  NProgress.isStarted = function() {
    return typeof status === 'number';
  };

  /**
   * Shows the progress bar.
   * This is the same as setting the status to 0%, except that it doesn't go backwards.
   *
   *     NProgress.start();
   *
   */
  NProgress.start = function() {
    if (!status) NProgress.set(0);

    var work = function() {
      setTimeout(function() {
        if (!status) return;
        NProgress.inc();
        work();
      }, 200);
    };

    work();

    return this;
  };

  /**
   * Hides the progress bar.
   * This is the *sort of* the same as setting the status to 100%, with the
   * difference being `done()` makes some placebo effect of some realistic motion.
   *
   *     NProgress.done();
   *
   * If `true` is passed, it will show the progress bar even if its hidden.
   *
   *     NProgress.done(true);
   */

  NProgress.done = function(force) {
    if (!force && !status) return this;

    return NProgress.inc(0.3 + 0.5 * Math.random()).set(1);
  };

  /**
   * Increments by a random amount.
   */

  NProgress.inc = function(amount) {
    var n = status;

    if (!n) {
      return NProgress.start();
    } else if(n > 1) {
      return;
    } else {
      if (typeof amount !== 'number') {
        if (n >= 0 && n < 0.2) { amount = 0.1; }
        else if (n >= 0.2 && n < 0.5) { amount = 0.04; }
        else if (n >= 0.5 && n < 0.8) { amount = 0.02; }
        else if (n >= 0.8 && n < 0.99) { amount = 0.005; }
        else { amount = 0; }
      }

      n = clamp(n + amount, 0, 0.994);
      return NProgress.set(n);
    }
  };

  /**
   * (Internal) renders the progress bar markup based on the `template`
   * setting.
   */

  NProgress.render = function(fromStart) {
    if (NProgress.isRendered()) return document.getElementById('nprogress');

    var progress = document.createElement('div');
    progress.id = 'nprogress';
    progress.innerHTML = '<div class="bar"><div class="peg"></div></div>';

    var bar      = progress.querySelector('#nprogress .bar'),
        perc     = fromStart ? '-100' : toBarPerc(status || 0);

    css(bar, {
      transition: 'all 0 linear',
      transform: 'translate3d(' + perc + '%,0,0)'
    });

    document.body.appendChild(progress);
    return progress;
  };

  /**
   * Removes the element. Opposite of render().
   */

  NProgress.remove = function() {
    var progress = document.getElementById('nprogress');
    progress && removeElement(progress);
  };

  /**
   * Checks if the progress bar is rendered.
   */

  NProgress.isRendered = function() {
    return !!document.getElementById('nprogress');
  };

  /**
   * Determine which positioning CSS rule to use.
   */

  function getPositioningCSS() {
    // Sniff on document.body.style
    var bodyStyle = document.body.style;

    // Sniff prefixes
    var vendorPrefix = ('WebkitTransform' in bodyStyle) ? 'Webkit' :
                       ('MozTransform' in bodyStyle) ? 'Moz' :
                       ('msTransform' in bodyStyle) ? 'ms' :
                       ('OTransform' in bodyStyle) ? 'O' : '';

    if (vendorPrefix + 'Perspective' in bodyStyle) {
      // Modern browsers with 3D support, e.g. Webkit, IE10
      return 'translate3d';
    } else if (vendorPrefix + 'Transform' in bodyStyle) {
      // Browsers without 3D support, e.g. IE9
      return 'translate';
    } else {
      // Browsers without translate() support, e.g. IE7-8
      return 'margin';
    }
  }

  /**
   * Helpers
   */

  function clamp(n, min, max) {
    if (n < min) return min;
    if (n > max) return max;
    return n;
  }

  /**
   * (Internal) converts a percentage (`0..1`) to a bar translateX
   * percentage (`-100%..0%`).
   */

  function toBarPerc(n) {
    return (-1 + n) * 100;
  }


  /**
   * (Internal) returns the correct CSS for changing the bar's
   * position given an n percentage and speed.
   */

  function barPositionCSS(n, speed) {
    var barCSS;

    if (positionUsing === 'translate3d') {
      barCSS = { transform: 'translate3d('+toBarPerc(n)+'%,0,0)' };
    } else if (positionUsing === 'translate') {
      barCSS = { transform: 'translate('+toBarPerc(n)+'%,0)' };
    } else {
      barCSS = { 'margin-left': toBarPerc(n)+'%' };
    }

    barCSS.transition = 'all '+speed+'ms linear';

    return barCSS;
  }

  /**
   * (Internal) Queues a function to be executed.
   */

  var queue = (function() {
    var pending = [];

    function next() {
      var fn = pending.shift();
      if (fn) {
        fn(next);
      }
    }

    return function(fn) {
      pending.push(fn);
      if (pending.length == 1) next();
    };
  })();

  /**
   * (Internal) Applies css properties to an element, similar to the jQuery
   * css method.
   *
   * While this helper does assist with vendor prefixed property names, it
   * does not perform any manipulation of values prior to setting styles.
   */

  var css = (function() {
    var cssPrefixes = [ 'Webkit', 'O', 'Moz', 'ms' ],
        cssProps    = {};

    function camelCase(string) {
      return string.replace(/^-ms-/, 'ms-').replace(/-([\da-z])/gi, function(match, letter) {
        return letter.toUpperCase();
      });
    }

    function getVendorProp(name) {
      var style = document.body.style;
      if (name in style) return name;

      var i = cssPrefixes.length,
          capName = name.charAt(0).toUpperCase() + name.slice(1),
          vendorName;
      while (i--) {
        vendorName = cssPrefixes[i] + capName;
        if (vendorName in style) return vendorName;
      }

      return name;
    }

    function getStyleProp(name) {
      name = camelCase(name);
      return cssProps[name] || (cssProps[name] = getVendorProp(name));
    }

    function applyCss(element, prop, value) {
      prop = getStyleProp(prop);
      element.style[prop] = value;
    }

    return function(element, properties) {
      var args = arguments,
          prop,
          value;

      if (args.length == 2) {
        for (prop in properties) {
          value = properties[prop];
          if (value !== undefined && properties.hasOwnProperty(prop)) applyCss(element, prop, value);
        }
      } else {
        applyCss(element, args[1], args[2]);
      }
    }
  })();

  /**
   * (Internal) Removes an element from the DOM.
   */

  function removeElement(element) {
    element && element.parentNode && element.parentNode.removeChild(element);
  }

  return NProgress;
})();
